java/src/security/Recursion/Recursion.ql

```diff
@@ -0,0 +1,93 @@
+/**
+ * @name Recursive functions
+ * @id tob/java/unbounded-recursion
+ * @description Detects possibly unbounded recursive calls
+ * @kind path-problem
+ * @tags security
+ * @precision low
+ * @problem.severity warning
+ * @security-severity 3.0
+ * @group security
+ */
+
+import java
+import semmle.code.java.dataflow.DataFlow
+
+predicate isTestPackage(RefType referenceType) {
+  referenceType.getPackage().getName().toLowerCase().matches("%test%") or
+  referenceType.getPackage().getName().toLowerCase().matches("%benchmark%") or
+  referenceType.getName().toLowerCase().matches("%test%")
+}
+
+class RecursionSource extends MethodCall {
+  RecursionSource() { not isTestPackage(this.getCaller().getDeclaringType()) }
+
+  override string toString() {
+    result = this.getCaller().toString() + " calls " + this.getCallee().toString()
+  }
+}
+
+/**
+ * Check if the Expr uses directly an argument of the enclosing function
+ */
+class ParameterOperation extends Expr {
+  ParameterOperation() {
```

Comment:
We may need parentheses here; now it reads:
```
(this instanceof BinaryExpr)
or
(
this instanceof UnaryAssignExpr
and
exists(...)
)
```

Reply:
Great catch!